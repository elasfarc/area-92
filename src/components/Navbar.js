import React from "react";
import PropTypes from "prop-types";
import { useHistory } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { library } from "@fortawesome/fontawesome-svg-core";

import { faAngleLeft } from "@fortawesome/free-solid-svg-icons";

library.add(faAngleLeft);

const Navbar = ({ children }) => {
  const history = useHistory();

  return (
    <div className="navbar mb-1">
      <div className="container  flex main-center cross-center">
        <button
          type="button"
          className="btn navbar__btn"
          onClick={() => history.push("/")}
        >
          <FontAwesomeIcon icon={faAngleLeft} />
        </button>
        {children}
      </div>
    </div>
  );
};

export default Navbar;

Navbar.propTypes = {
  children: PropTypes.node,
};

Navbar.defaultProps = {
  children: null,
};
