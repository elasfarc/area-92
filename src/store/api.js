// ACTION TYPES
export const API_CALL_REQUEST = "/api/request";
export const API_CALL_SUCCESS = "/api/success";
export const API_CALL_FAIL = "/api/fail";

// ACTION CREATOR
export const requestApiCall = (payload) => ({
  type: API_CALL_REQUEST,
  payload,
});
export const onApiSuccess = () => ({ type: API_CALL_SUCCESS });
export const onApiFail = () => ({ type: API_CALL_FAIL });
