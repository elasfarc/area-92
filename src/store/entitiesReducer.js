import { combineReducers } from 'redux'
import countryReducer from './country'

export default  combineReducers({
  countries: countryReducer
})