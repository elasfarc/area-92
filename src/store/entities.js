import * as apiActions from "./api";

// ACTION TYPES
const COUNTRIES_REQUESTED = "/countries/request";
const COUNTRIES_LOADED = "/countries/load";
const STATES_LOADED = "/country/states";
const CITIES_TRANSFORM_TO_GEO = "country/cities/transform2Geo";
const CITY_WEATHER_LOADED = "/weather/city";

// async thunk
export const loadCountriesData = () => async (dispatch) => {
  dispatch({ type: COUNTRIES_REQUESTED });
  try {
    const { data: countries } = await (
      await fetch(
        `https://countriesnow.space/api/v0.1/countries/info?returns=flag,currency,cities`
      )
    ).json();
    const { data: capitals } = await (
      await fetch(`https://countriesnow.space/api/v0.1/countries/capital`)
    ).json();
    dispatch(apiActions.onApiSuccess());
    dispatch({
      type: COUNTRIES_LOADED,
      payload: { countries, capitals },
    });
  } catch (error) {
    dispatch(apiActions.onApiFail());
  }
};

// ACTION CREATOR

export const getCountryStates = (country) =>
  apiActions.requestApiCall({
    url: `https://countriesnow.space/api/v0.1/countries/states`,
    method: "POST",
    body: { country },
    onStart: COUNTRIES_REQUESTED,
    onSuccess: STATES_LOADED,
    info: country,
  });

export const transformCitiesToGeo = ({ country, cities }) =>
  apiActions.requestApiCall({
    url: `http://open.mapquestapi.com/geocoding/v1/batch?key=4qiFlGPdSaIgG97k8C3AyJRKLBpq1jtJ`,
    method: "POST",
    body: { locations: cities.map((city) => ({ city, country })) },
    onStart: COUNTRIES_REQUESTED,
    onSuccess: CITIES_TRANSFORM_TO_GEO,
    info: country,
  });

export const getGeoWeather = () => ({});

// REDUCER
const initialState = {
  countries: [],
  isLoading: false,
};

const reducer = (state = initialState, { type, payload }) => {
  if (type === COUNTRIES_REQUESTED) return { ...state, isLoading: true };

  if (type === COUNTRIES_LOADED) {
    const { countries, capitals } = payload;
    return {
      ...state,
      isLoading: false,
      countries: countries.map((country) => ({
        ...country,
        ...capitals.find(({ capital, name }) => country.name === name),
      })),
    };
  }

  if (type === STATES_LOADED) {
    const {
      response: {
        data: { states },
      },
      info: countryName,
    } = payload;

    return {
      ...state,
      isLoading: false,
      countries: state.countries.map((country) =>
        country.name === countryName ? { ...country, states } : country
      ),
    };
  }

  if (type === CITIES_TRANSFORM_TO_GEO) {
    return {
      ...state,
      isLoading: false,
      countries: state.countries.map((country) =>
        country.name === payload.info
          ? {
              ...country,
              cities: payload.response.results.map(
                ({
                  providedLocation,
                  locations: [{ latLng, mapUrl, adminArea3 }],
                }) => ({
                  name: providedLocation.city,
                  country: country.name,
                  state: adminArea3,
                  latLng,
                  mapUrl,
                })
              ),
            }
          : country
      ),
    };
  }
  if (type === CITY_WEATHER_LOADED) return state;
  return state;
};

export default reducer;
