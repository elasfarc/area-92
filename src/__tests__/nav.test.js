import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import Navbar from "../components/Navbar";
import store from "../store/configureStore";

const MockNav = () => (
  <Provider store={store}>
    <Navbar />
  </Provider>
);

describe("Mission component", () => {
  it("should match the snapshot", () => {
    const tree = renderer.create(<MockNav />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
