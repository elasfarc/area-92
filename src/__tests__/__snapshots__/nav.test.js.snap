// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission component should match the snapshot 1`] = `
<div
  className="navbar mb-1"
>
  <div
    className="container  flex"
  >
    <div
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-left fa-w-8 "
        data-icon="angle-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
